drain manager: Refactor DrainManagerImpl with simulated time

auni53 opened the pull request:
drain_manager: Refactor DrainManagerImpl with simulated time
\* Replace recurring timer in DrainManagerImpl with a single timer that can be tested via simulated time. 
\* Clarify the behaviour where probability of drain close increases as the deadline is approached, and add mocks on the random() call to test this
\* Rely on the system scheduler rather than tick time
\* Had to remove the "drain tick #x" trace log
\* Remove overly explicit includes
\* Add TODO for integration tests on this via a real time system, once TestDrainManager is removed (that's a separate thread of work)

Additional Description: Preparatory work for enhancing graceful drain functionality.
Risk Level: Medium. This is changing a core dependency (via the server) that hasn't been touched in a while, but the behaviour should be a noop.
Testing: I added a lot more testing coverage, and some aspects of this (the probabilistic draining) weren't being tested at all prior.

mattklein123 commented on source/server/drain_manager_impl.cc:
```diff
@@ -35,30 +32,24 @@ bool DrainManagerImpl::drainClose() const {
-    drain_sequence_completion_();
-  }
+  // P(shut down connections) = elapsed_drain_time / overall_drain_timeout
+  return static_cast<uint64_t>(elapsed_time.count()) >
+         (server_.random().random() % server_.options().drainTime().count());
```
note that this has performance implications in terms of asking for a new random number for each drain decision. I'm not sure that this matters but it's worth calling out. 

Stepping back, what is the reasoning for this change? It's not clear to me why this needs to change.

auni53 replied:
I don't think the random() calls have changed, the existing code was already calling for a random() on each drain decision.

Reasoning for the change:

1. I'm planning to enhance integration testing coverage around the drain period given the additional complexity I'm planning to add. I'd like to reliably use simulated time to write those tests.

2. The current approach rearms the timer every second just for the purpose of updating the drain time completed, but this seems less precise since it has to run the code prior to the rearming on each callback. Having a single timer arming makes the behaviour easier to reason about in simulated time.

3. I found the existing code difficult to understand at a glance, and it wasn't sufficiently tested. Computing the deadline one time at the start simplifies the code a lot, I had a hard time understanding the whole drainSequenceTick thing. It's also easier to extend or modify, in case I end up needing to deal with overriding drain close periods or new phases in the drain sequence.

mattklein123 replied:
OK I missed the diff in which we did random in the old code. OK fair enough sounds fine to me.

auni53 commented:
/retest

jmarantz commented on source/server/drain_manager_impl.cc:
```diff
@@ -35,30 +32,25 @@ bool DrainManagerImpl::drainClose() const {
-  drain_tick_timer_ = server_.dispatcher().createTimer([this]() -> void { drainSequenceTick(); });
-  drainSequenceTick();
+  drain_tick_timer_ =
+      server_.dispatcher().createTimer(drain_complete_cb_ ? drain_complete_cb_ : [] {});
```
is there a prod scenario where drain_complete_cb is false? I see it in the tests, but you could just put the empty callback there, rather than adding to the complexity of the prod code.

auni53 replied:
The problem is actually with the timer API, which has an assertion on callbacks being non-false. I don't really think the check makes sense, since noop functions fail the assert but don't do anything when called, but I didn't think it made sense to mess with that in this PR.

jmarantz replied:
That's fine. Just assert that drain_callback_cb_ is truthy, and change the test to pass an empty callback rather than null.

auni53 replied:
To answer your first question, there actually is a prod scenario. server.cc:139 has a `drain_manager_->startDrainSequence(nullptr);`. I can just change that though?

The reason this is coming up now is because prior, the timer didn't receive the cb directly, but rather a lambda that invoked the cb. So I guess another option that's neater is `server_.dispatcher().createTimer([]() { drain_complete_cb(); }`.

jmarantz replied:
Yeah I think we should just change the callsite that passes null for the functor to simplify all the other code paths.

auni53 replied:
Done.

jmarantz commented on source/server/drain_manager_impl.cc:
```diff
@@ -35,30 +32,25 @@ bool DrainManagerImpl::drainClose() const {
-void DrainManagerImpl::startDrainSequence(std::function<void()> completion) {
-  drain_sequence_completion_ = completion;
+void DrainManagerImpl::startDrainSequence(std::function<void()> drain_complete_cb) {
+  drain_complete_cb_ = drain_complete_cb;
```
why do we need to hold this callback in the DrainManagerImpl? Isn't it enough to hold it in the timer?

auni53 replied:
I wasn't really sure about the lifetime parameters of `std::function`s but that's fine by me.

jmarantz commented on source/server/drain_manager_impl.cc:
```diff
@@ -35,30 +32,25 @@ bool DrainManagerImpl::drainClose() const {
-  ASSERT(drain_time_completed_.load() < server_.options().drainTime().count());
-  ++drain_time_completed_;
+  const auto remaining_time = std::chrono::duration_cast<std::chrono::seconds>(
+      drain_deadline_ - server_.dispatcher().timeSource().monotonicTime());
```
what if time has moved past the drain deadline?  You'll get a negative number from your subtraction. Will this code work?

Can you strobe the monotonicTime() and compare it to drain_deadline_ before subtracting?

auni53 replied:
I don't understand what you mean by strobing, but I added tests to cover the negative case. The subtraction works out, elapsed time will exceed the drain deadline, and therefore always return true regardless of the `random() % drain_timeout`.

jmarantz replied:
strobing just means to capture the current monotonicTime() once in a temp, rather than calling it twice (once for the compare and once for the subtract), with its value potentially changing between the two calls.

I was searching for a EE reference for that term but can't find the usage I always thought of, which was analogous to a strobe in photography where you can capture a moment in time even with a fast moving object.

auni53 replied:
Done.

\* Put the current time into a variable
\* Return true if current time is past the deadline
\* Extra testing to ensure random() is only called when elapsed time is less the drain timeout, even if the random() returned would cause drainClose() to return true

jmarantz commented:
/wait

auni53 commented:
Sorry for the format presubmit fails...I thought I had it setup to automatically lint before pushing, trying to fix that.

jmarantz commented:
/wait

auni53 commented:
/retest

auni53 commented:
> 🐴 hold your horses - no failures detected, yet.
> 
> 🐱

auni53 commented:
/retest

auni53 commented:
I don't understand why these presubmits are failing...unable to reproduce yet. Going to merge from master and push up again.

jmarantz commented on test/server/drain_manager_impl_test.cc:
```diff
@@ -17,10 +17,13 @@ namespace Envoy {
 namespace {
 
-class DrainManagerImplTest : public testing::Test {
+const int kDrainTimeSeconds(600);
```
`constexpr int DrainTickSeconds = 600;`

Envoy naming convention is slightly different from Google's.

auni53 replied:
Done, but can you explain what is meant by "tick" here as compared to time? I understand CPU clocks ticking and stuff, but is it just commonplace to refer to time as ticks?

jmarantz replied:
I have no idea why I wrote 'Tick'. I meant DrainTimeSeconds.

auni53 commented:
Fix the naming nit, is this good to go?

jmarantz commented:
@envoyproxy/senior-maintainers

mattklein123 commented on source/server/drain_manager_impl.cc:
```diff
@@ -35,30 +32,29 @@ bool DrainManagerImpl::drainClose() const {
-  drain_tick_timer_ = server_.dispatcher().createTimer([this]() -> void { drainSequenceTick(); });
-  drainSequenceTick();
+  drain_tick_timer_ =
+      server_.dispatcher().createTimer(drain_complete_cb ? drain_complete_cb : [] {});
```
You assert drain_complete_cb above so I don't think this will ever execute the other branch?

auni53 commented:
(I'm waiting to resolve #11240 before introducing more changes.)

auni53 commented:
Fixed the ternary, added an expected_delay to the mock expect, and put up #11345 which addresses the tangential race condition. Hope this is good to go!

auni53 commented:
/retest

jmarantz commented on source/server/drain_manager_impl.cc:
```diff
@@ -35,30 +32,28 @@ bool DrainManagerImpl::drainClose() const {
+  // P(return true) = elapsed time / drain timeout
+  const auto remaining_time =
+      std::chrono::duration_cast<std::chrono::seconds>(drain_deadline_ - current_time);
+  const auto elapsed_time = server_.options().drainTime() - remaining_time;
```
is there a guarantee that drainTime() >= current_time ? Or should we do a compare first? I don't totally understand the surrounding draining architecture, but it looks like drainClose() can be called by something not controlled by the timer, so current_time_ may just be later. In that case maybe we should skip the subtraction underflow and compare to random and just return true?

auni53 replied:
Assuming you mean drain_deadline_ (the computed deadline time point) and not drainTime() (the server configuration)? This case is tested for at [drain_manager_impl_test.cc:83](https://github.com/envoyproxy/envoy/pull/11138/files/3edaebebe35821a32b9a7b8b2112c3e0729633c9#diff-8878b807be959b98509feb61ed0452c7R82). If current_time > drain_deadline, the probability of drain close is 100%,

jmarantz replied:
sorry I meant checking server_.options().drainTime() >= remaining_time

or is that guaranteed to be true due to the wait remaining_time is calculated?

auni53 replied:
(psuedo variables)

drain_deadline = drain_timeout + previous_time
remaining_time = drain_deadline - current_time
remaining_time = drain_timeout + previous_time - current_time
current_time >= previous time, therefore remaining_time <= drain_timeout

Math checks out to me!

jmarantz replied:
awesome. add assert then?

auni53 replied:
Not sure if it's the exact ASSERT you wanted but 

drain >= remaining
drain - remaining >= 0
elapsed = drain - remaining
elapsed >= 0

so I added ASSERT(elapsed_time >= 0);

jmarantz replied:
probably not, if elapsed_time is ultimately a uint_something.

I meant asserting the subtraction is non-underflowing.
```
ASSERT(server_.options().drainTime() >= remaining_time);
```

auni53 commented:
@mattklein123 is this good to merge?

mattklein123 commented on test/server/drain_manager_impl_test.cc:
```diff
@@ -17,10 +17,13 @@ namespace Envoy {
 namespace {
 
-class DrainManagerImplTest : public testing::Test {
+constexpr int DrainTimeSeconds(600);
```
nit: just use std::chrono::seconds here. Feel free to do this change in one of your other PRs.
